src/NServiceBus.TransactionalSession.Tests/ApprovalFiles/APIApprovals.Approve.approved.txt

```diff
@@ -24,6 +24,7 @@ namespace NServiceBus.TransactionalSession
         protected NServiceBus.Extensibility.ContextBag Extensions { get; }
         public System.TimeSpan MaximumCommitDuration { get; set; }
         public System.Collections.Generic.IDictionary<string, string> Metadata { get; }
+        public bool OmitOutboxRecordWhenThereAreNoOutgoingOperations { get; set; }
```

Comment:
Maybe more explicit about the storage, but I'd be fine with what you have too.

```suggestion
        public bool DoNotStoreOutboxRecordWhenThereAreNoOutgoingOperations { get; set; }
```

Reply:
Maybe it is clearer to use a non-negated naming and flip the default. Something like

`StoreOutboxRecordsEvenWhenThereAreNoOutgoingOperations { get; set; } = true;`


The Microsoft Framework Design Guidelines suggest the following:

- Use positive Boolean property names: Avoid naming Boolean properties with a negative tone. It’s clearer to have a property like IsEnabled instead of IsNotDisabled.
- Keep names simple and consistent: Boolean properties should be easily understandable and concise, ideally reflecting a clear true/false outcome without confusion.

or the super short form

AlwaysStoreOutboxRecords

but I wonder if that could be confusing due to its broader definition

Reply:
Agreed on the non-negative thing. What about something like `StoreOutboxRecordsForEmptyOutgoingOperations` or `StoreEmptyOutgoingOperationsOutboxRecords`?

Reply:
Making a step back seeing all these ideas I think we might be overly focuses on the technical aspects of the storing. But when you look at what the outbox does is that eventually when the messages (or no messages) are dispatched it removes the outgoing operations and sets the record as dispatched. From that point in time the record serves as a pure deduplication entry representing the ID of the incoming messages and a time (more or less). 

So in essence, our default behavior here does the same => Creating the deduplication entry by storing the record and marking it as dispatched. So the optimization here is not opting out from storing outgoing operations but essentially opting-out from keeping the deduplication record. Would you agree to that?

So how about

- AlwaysStoreDeduplicationRecord
- StoreDeduplicationRecord
- KeepDeduplicationRecord

or something of the like?

Reply:
I find it a bit confusing that we introduce an API that enables behavior that is conditional: the `Always*` prefix is sort of a smell to me because what is the opposite? Sometimes? i.e. we could name it to `OnlySometimesKeepDeduplicationRecords`.

I'd rather wait until we expose the transaction ID and then allow two definitive modes: with and without deduplication. If you provide the ID, we always created a record (even if empty). If you don't provide the ID, we get a random one and we _may_ create a record under that random one, but you have no way to retrieve it because we will not let you see that random ID.

Reply:
We are thinking about making this opt-in because we have real feedback from users that in their system the have large amounts of conditional that leads to many operations not sending out anything. The IO these Inserts and Updates still creates is non-trivial and has a real impact

The thought process that if you know the implications and you don´t need to keep the dedup record you can opt out

I get your point about the user provided ID and can see how those two things could be related, but it is also possible that they are not (we just don't know yet)

So I was wondering if the decision of exposing that flag could not be reversed later once we expose the session ID

for example, it could also be possible to validate that you can't have that flag set and the session set or once we expose the session ID we could deprecate the other property and make the behavior you described standard

which makes me think it is a low-risk decision and is easily reversible to expose some kind of flag

Reply:
After discussions with @SzymonPobiega I raised https://github.com/Particular/NServiceBus.TransactionalSession/pull/352 and think this should be closed. Waiting for @mauroservienti and @SzymonPobiega to chime in there.